### PR TITLE
Fixed "False error when using directive prologues inside constructor with initialized properties."

### PR DESCRIPTION
**Pull request by ShyykoSerhiy**
Fixes issue #3859

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10270,12 +10271,22 @@ namespace ts {
 
                     if (superCallShouldBeFirst) {
                         let statements = (<Block>node.body).statements;
-                        if (!statements.length || statements[0].kind !== SyntaxKind.ExpressionStatement || !isSuperCallExpression((<ExpressionStatement>statements[0]).expression)) {
-                            error(node, Diagnostics.A_super_call_must_be_the_first_statement_in_the_constructor_when_a_class_contains_initialized_properties_or_has_parameter_properties);
+                        let superCallStatement: ExpressionStatement;
+                        for (let i = 0; i < statements.length; i++) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use a for-of loop

```TypeScript
for (let statement of statements) {
}
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10270,12 +10271,22 @@ namespace ts {
...
+                                break;
+                            }
+                            if (!isPrologueDirective(statement)) {
+                                break;
+                            }
                         }
-                        else {
+                        if (!superCallStatement) {
+                            error(node, Diagnostics.A_super_call_must_be_the_first_statement_in_the_constructor_when_a_class_contains_initialized_properties_or_has_parameter_properties);
+                        } else {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `else` on the next line

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -10270,12 +10271,22 @@ namespace ts {
 
                     if (superCallShouldBeFirst) {
                         let statements = (<Block>node.body).statements;
-                        if (!statements.length || statements[0].kind !== SyntaxKind.ExpressionStatement || !isSuperCallExpression((<ExpressionStatement>statements[0]).expression)) {
-                            error(node, Diagnostics.A_super_call_must_be_the_first_statement_in_the_constructor_when_a_class_contains_initialized_properties_or_has_parameter_properties);
+                        let superCallStatement: ExpressionStatement;
+                        for (let statement of statements) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Let's add a comment above:

```TypeScript
// Skip past prologue directives to find the first statement
// to ensure that it was a super call.
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -10260,7 +10260,8 @@ namespace ts {
             if (getClassExtendsHeritageClauseElement(<ClassDeclaration>node.parent)) {
 
                 if (containsSuperCall(node.body)) {
-                    // The first statement in the body of a constructor must be a super call if both of the following are true:
+                    // The first statement in the body of a constructor must be a super call(or prologue directives followed by a super call)
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Maybe just turn this into

```TypeScript
// The first statement in the body of a constructor (excluding prologue directives) must be a super call.
```

If you prefer your original wording, just split the comment into two linese.

**Comment by mhegazy** [comment]
Thanks!

**Pull request merged**